src/Service/TokenService.cs

```diff
@@ -237,4 +207,39 @@ private bool VerifyMac(Key key, byte[] data, byte[] mac)
 
         return result;
     }
+
+    private async Task<Dictionary<int, string>> GetAlternativesAsync()
+    {
+        if (_alternatives == null)
+        {
+            var keys = await _storage.GetTokenKeys();
+
+            // transform our list to a dictionary
+            _alternatives = keys.OrderByDescending(x => x.CreatedAt).ToDictionary(k => k.KeyId, k => k.KeyMaterial);
+
+            // Rotate keys every 7 day
+            // Remove old keys after 30 days (side effect: Tokens maximum life length is 30 days).
+            if (!_alternatives.Any() || (DateTime.UtcNow - keys.First().CreatedAt).TotalDays > 7)
+            {
+                var random32Bytes = RandomNumberGenerator.GetBytes(32);
+                var keyInputMaterial = Convert.ToBase64String(random32Bytes);
+
+                // todo: Handle 409 exception from storage? Storage will throw if keyid is duplicate.
+                var keyId = RandomNumberGenerator.GetInt32(int.MaxValue);
+                await _storage.AddTokenKey(new TokenKey { CreatedAt = DateTime.UtcNow, KeyId = keyId, KeyMaterial = keyInputMaterial });
+                _alternatives.Add(keyId, keyInputMaterial);
+
+                try
+                {
+                    await _storage.RemoveExpiredTokenKeys(CancellationToken.None);
+                }
+                catch (Exception)
+                {
+                    _log.LogError("Failed to remove old key, account={accountName}", _tenant);
+                }
+            }
+        }
+
+        return _alternatives;
```

Comment:
There's a potential of a race condition if `GetAlternativeAsync` is called twice in quick succession while `_alternatives` is null.

It might be better to initialize a local `var alternatives`, perform all manipulations, and only set `_alternative = alternatives` in the return statement